### PR TITLE
Add option to hide bottom view controller

### PR DESCRIPTION
**Pull request by felixLam**

### CODE REVIEW

File: ISHPullUp/ISHPullUpViewController.h [hunk 1]
```diff
@@ -190,6 +190,20 @@ typedef struct ISHPullUpAnimationConfiguration ISHPullUpAnimationConfiguration;
 /// The layout mode to be used for the bottom view controller. Default is ISHPullUpBottomLayoutModeShift.
 @property (nonatomic) ISHPullUpBottomLayoutMode bottomLayoutMode;
 
+/// YES if bottom is hidden outside of view
+@property (nonatomic) BOOL bottomHidden;
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.h]: Maybe make this `readonly` to make `setBottomHidden:animated:` more discoverable?

File: ISHPullUp/ISHPullUpViewController.h [hunk 2]
```diff
@@ -190,6 +190,20 @@ typedef struct ISHPullUpAnimationConfiguration ISHPullUpAnimationConfiguration;
...
+
+/// Hide or unhide the bottom view controller.
+- (void)setBottomHidden:(BOOL)bottomHidden animated:(BOOL)animated;
+
+/**
+ *  When hiding the bottom view this additional margin is added to allow 
+ *  shadows and other elements that extend beyond the bottom viewcontroller's
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.h]: `view controller's`

File: ISHPullUp/ISHPullUpViewController.h [hunk 3]
```diff
@@ -190,6 +190,20 @@ typedef struct ISHPullUpAnimationConfiguration ISHPullUpAnimationConfiguration;
...
+/// YES if bottom is hidden outside of view
+@property (nonatomic) BOOL bottomHidden;
+
+/// Hide or unhide the bottom view controller.
+- (void)setBottomHidden:(BOOL)bottomHidden animated:(BOOL)animated;
+
+/**
+ *  When hiding the bottom view this additional margin is added to allow 
+ *  shadows and other elements that extend beyond the bottom viewcontroller's 
+ *  view to be completly hidden.
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.h]: `completely`

File: ISHPullUp/ISHPullUpViewController.m [hunk 4]
```diff
@@ -373,16 +378,77 @@ - (void)updateCachedHeightsWithSize:(CGSize)size  {
     self.maximumBottomHeightCached = [self maximumBottomHeightWithSize:size];
 }
 
+- (void)setBottomHidden:(BOOL)bottomHidden {
+    [self setBottomHidden:bottomHidden animated:NO];
+}
+
+- (void)setBottomHidden:(BOOL)bottomHidden animated:(BOOL)animated {
+    if (bottomHidden == _bottomHidden) {
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.m]: `!bottomHidden == !_bottomHidden`

File: ISHPullUp/ISHPullUpViewController.m [hunk 5]
```diff
@@ -373,16 +378,77 @@ - (void)updateCachedHeightsWithSize:(CGSize)size  {
...
+
+    if (!bottomHidden) {
+        bottomVC.view.hidden = NO;
+        [bottomVC viewWillAppear:animated];
+    } else {
+        [bottomVC viewWillDisappear:animated];
+    }
+
+    void (^updateBlock)();
+    updateBlock = ^{
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.m]: Why not on one line?

File: ISHPullUp/ISHPullUpViewController.m [hunk 6]
```diff
@@ -373,16 +378,77 @@ - (void)updateCachedHeightsWithSize:(CGSize)size  {
...
+        [self setDimmingViewHidden:[self dimmingViewShouldBeHidden] height:self.bottomHeight];
+        [self updateViewLayoutBottomHeight:self.bottomHeight withSize:self.view.bounds.size];
+    };
+
+    void (^completion)(BOOL);
+    completion = ^(BOOL finished) {
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.m]: Why not on one line?

File: ISHPullUp/ISHPullUpViewController.m [hunk 7]
```diff
@@ -373,16 +378,77 @@ - (void)updateCachedHeightsWithSize:(CGSize)size  {
...
+    if (!bottomHidden) {
+        bottomVC.view.hidden = NO;
+        [bottomVC viewWillAppear:animated];
+    } else {
+        [bottomVC viewWillDisappear:animated];
+    }
+
+    void (^updateBlock)();
+    updateBlock = ^{
+        [self setDimmingViewHidden:[self dimmingViewShouldBeHidden] height:self.bottomHeight];
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.m]: `weakSelf`

File: ISHPullUp/ISHPullUpViewController.m [hunk 8]
```diff
@@ -373,16 +378,77 @@ - (void)updateCachedHeightsWithSize:(CGSize)size  {
...
+    };
+
+    void (^completion)(BOOL);
+    completion = ^(BOOL finished) {
+        if (!finished) {
+            return;
+        }
+
+        if (bottomHidden) {
+            [bottomVC viewDidDisappear:animated];
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.m]: Should we rather use `weakSelf.bottomViewController` (get it once within the block, return early if `nil`) to ensure:

1. ...we get the current bottom vc
2. ...we don't hold on to the bottom vc if `self` has been released?

File: ISHPullUp/ISHPullUpViewController.m [hunk 9]
```diff
@@ -404,19 +470,28 @@ - (void)setBottomHeight:(CGFloat)bottomHeight animated:(BOOL)animated {
         return;
     }
 
-    ISHPullUpAnimationConfiguration config = self.animationConfiguration;
-    NSAssert(config.options & UIViewAnimationOptionLayoutSubviews, @"Animation options must always contain UIViewAnimationOptionLayoutSubviews");
+    [[self class] springAnimationWithConfig:self.animationConfiguration
+                                 animations:updateBlock
+                                 completion:^(BOOL finished) {
+                                     self.isAnimatingStateChange = NO;
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.m]: `weakSelf`, and maybe even transform back to `strongSelf` within block

File: ISHPullUp/ISHPullUpViewController.m [hunk 10]
```diff
@@ -456,6 +531,11 @@ - (void)updateViewLayoutBottomHeight:(CGFloat)bottomHeight withSize:(CGSize)size
         }
     }
 
+    if (self.bottomHidden) {
+        // hide the bottom view by moving the view below outside the view and add the bottomHiddenMargin
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.m]: ~~outside~~

File: ISHPullUp/ISHPullUpViewController.m (hunk 7 above)
Comment (felixLam) [inline ISHPullUp/ISHPullUpViewController.m]: The update block is executed immediately, I will keep as is as we usually skip this extra layer for animation blocks

File: ISHPullUp/ISHPullUpViewController.m [hunk 11]
```diff
@@ -383,7 +384,7 @@ - (void)setBottomHidden:(BOOL)bottomHidden {
 }
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.m]: Remove this stub, too? ^

File: ISHPullUp/ISHPullUpViewController.m [hunk 12]
```diff
@@ -399,23 +400,23 @@ - (void)setBottomHidden:(BOOL)bottomHidden animated:(BOOL)animated {
...
-    updateBlock = ^{
+    void (^updateBlock)() = ^{
         [self setDimmingViewHidden:[self dimmingViewShouldBeHidden] height:self.bottomHeight];
         [self updateViewLayoutBottomHeight:self.bottomHeight withSize:self.view.bounds.size];
     };
 
-    void (^completion)(BOOL);
-    completion = ^(BOOL finished) {
-        if (!finished) {
+    __weak typeof(self) weakSelf = self;
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.m]: No need for `typeof(self)`... just use `ISHPullUpViewController  *`

File: ISHPullUp/ISHPullUpViewController.m [hunk 13]
```diff
@@ -399,23 +400,23 @@ - (void)setBottomHidden:(BOOL)bottomHidden animated:(BOOL)animated {
...
+    void (^completion)(BOOL) = ^(BOOL finished) {
+        __strong typeof(self) strongSelf = weakSelf;
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.m]: No need for `typeof(self)`... just use `ISHPullUpViewController  *`. `__strong` is probably not needed either, as that's the default for local variables.

File: ISHPullUp/ISHPullUpViewController.m [hunk 14]
```diff
@@ -470,11 +471,16 @@ - (void)setBottomHeight:(CGFloat)bottomHeight animated:(BOOL)animated {
         return;
     }
 
+    __weak typeof(self) weakSelf = self;
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.m]: See above

File: ISHPullUp/ISHPullUpViewController.m [hunk 15]
```diff
@@ -470,11 +471,16 @@ - (void)setBottomHeight:(CGFloat)bottomHeight animated:(BOOL)animated {
...
     [[self class] springAnimationWithConfig:self.animationConfiguration
                                  animations:updateBlock
                                  completion:^(BOOL finished) {
-                                     self.isAnimatingStateChange = NO;
-                                     [self.stateDelegate pullUpViewController:self didChangeToState:[self state]];
+                                     __strong typeof(self) strongSelf = weakSelf;
```
Comment (shagedorn) [inline ISHPullUp/ISHPullUpViewController.m]: see above

**Pull request merged**